runtime/src/main/java/org/corfudb/runtime/view/replication/ChainReplicationProtocol.java

```diff
@@ -73,96 +73,140 @@ public ILogData peek(RuntimeLayout runtimeLayout, long globalAddress) {
                         .getLogUnitClient(globalAddress, numUnits - 1)
                         .read(globalAddress)).getAddresses()
                 .getOrDefault(globalAddress, null);
-        return ret == null || ret.isEmpty() ? null : ret;
+        return !isLogDataPresent(ret) ? null : ret;
     }
 
     /**
-     * {@inheritDoc}
+     * Reads a list of global addresses from the chain of log unit servers.
+     *
+     * - This method optimizes for the time to wait to hole fill in case an empty address
+     * is encountered.
+     * - If the waitForWrite flag is set to true, when an empty address is encountered,
+     * it waits for the hole to be filled. All subsequent empty addresses are hole filled
+     * directly and the reader does not wait.
+     * - In case the flag is set to false, none of the reads wait for write completion and
+     * the empty addresses are hole filled right away.
+     *
+     * @param runtimeLayout runtime layout.
+     * @param addresses     list of addresses to read.
+     * @param waitForWrite  flag whether wait for write is required or hole fill directly.
+     * @return Map of read addresses.
      */
     @Override
-    public Map<Long, ILogData> readAll(RuntimeLayout runtimeLayout, List<Long> globalAddresses) {
-        Range<Long> range = Range.encloseAll(globalAddresses);
-        return readRange(runtimeLayout, range, true);
-    }
-
+    @Nonnull
+    public Map<Long, ILogData> readAll(RuntimeLayout runtimeLayout,
+                                       List<Long> addresses,
+                                       boolean waitForWrite) {
+        long startAddress = Collections.min(addresses);
+        long endAddress = Collections.max(addresses);
+        int numUnits = runtimeLayout.getLayout().getSegmentLength(startAddress);
+        log.trace("readAll[{}-{}]: chain {}/{}", startAddress, endAddress, numUnits, numUnits);
 
+        Map<Long, LogData> logResult = CFUtils.getUninterruptibly(runtimeLayout
```

Comment:
Maybe `readResult` is a better variable name

Reply:
Done.